sourcecred/sourcecred
# Convert Explorer Home timeline summaries to for loop for greater efficiency

# Description
In tidying up the Explorer Home page for launch, I noticed a bit of lag upon switching the timeline control, especially for longer periods (e.g. All Time). I profiled the code and saw that compiling the timeline data is fairly intensive. This is understandable given that we're looping through the array of all participants, and within that looping though all time periods and summing the info.

In order to speed this up, I refactored the maps and reduces into one pass for both cred and grain summaries. This sped things up around 20% for All Time calculations. Below are timings in milliseconds for This Week, Last Week, This Month, All Time in order.

### Original timings in ms
![Screen Shot 2021-03-02 at 22 58 44](https://user-images.githubusercontent.com/225508/110551289-889a5e80-8135-11eb-8cee-1cb43dd4b654.png)


### Refactored timings in ms
![Screen Shot 2021-03-02 at 23 36 05](https://user-images.githubusercontent.com/225508/110551300-8cc67c00-8135-11eb-882a-a7d530367e45.png)

# Test Plan
Tested manually with the following plan:


* Load page
* Click on different time periods
* Ensure that the top graph changes accordingly



┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200032326400385/1200040151419673) by [Unito](https://www.unito.io/learn-more)
